### ISSUE TITLE
Improve dependency injection registration

### ISSUE THREAD

**Issue by JamesNK** [open-telemetry/opentelemetry-dotnet#241]
Currently OT documentation asks the user to registered a lot of services:

```cs
services.AddSingleton<ISampler>(Samplers.AlwaysSample);
services.AddSingleton<ZipkinTraceExporterOptions>(_ => new ZipkinTraceExporterOptions { ServiceName = "my-service" });
services.AddSingleton<SpanExporter, ZipkinTraceExporter>();
services.AddSingleton<SpanProcessor, BatchingSpanProcessor>();
services.AddSingleton<TraceConfig>();
services.AddSingleton<ITracer, Tracer>();

// you may also configure request and dependencies collectors
services.AddSingleton<RequestsCollectorOptions>(new RequestsCollectorOptions());
services.AddSingleton<RequestsCollector>();

services.AddSingleton<DependenciesCollectorOptions>(new DependenciesCollectorOptions());
services.AddSingleton<DependenciesCollector>();
```

This internal implementation detail should be hidden behind a helper method, e.g.

https://github.com/grpc/grpc-dotnet/blob/f448189939124bd38a4924754e0256b0da671391/src/Grpc.AspNetCore.Server/GrpcServiceExtensions.cs#L58

There should be a friendly user experience for OT configuration.

**Comment by lmolkova** [comment]
@JamesNK you are right, we should provide such a method.
We are now focused on delivering and stabilizing core components (Abstractions and implementation packages with a few exporters) and integration with ASP.NET Core will be the next step for us.

We are happy to get contributions :)

**Comment by JamesNK** [comment]
We'll see. At the very least I can help with design and review.

I think something like ConfigureLogging is what you would want to do - https://wakeupandcode.com/logging-in-asp-net-core/

e.g.

```cs
services.ConfigureOpenTelemetry(telemetry =>
{
    telemetry.AddZipkin("zipkin configuration args here");
});

**Comment by lmolkova** [comment]
Yeah, something similar would be great. 

We also have another example for [AppInsights](https://github.com/microsoft/ApplicationInsights-aspnetcore/blob/develop/src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs#L124) created initially by @pakrym 

There are couple of complications  I see
1. Exporters cannot have reference to Ms.Ext.DI and you cannot configure exporter together with the rest of OpenTelemetry. At the same time, OpenTelemtery.ASP.NET Core cannot know about all exporters

2. Auto-collectors (requests, dependency) today need to be resolved manually  - nothing depends on them and can initiate collection

**Comment by JamesNK** [comment]
> Exporters cannot have reference to Ms.Ext.DI

Exporters could each have an additional project that contains just an extension method for registering it with DI, e.g. `OpenTelemetry.Exporter.Zipkin.Abstractions`. The project would then reference DI and `OpenTelemetry.Exporter.Zipkin`.

> Auto-collectors (requests, dependency) today need to be resolved manually - nothing depends on them

I don't know of how to fix that off the top of my head. It would need some investigation. There is probably some nice way to improve it though.

**Comment by lmolkova** [comment]
> Exporters could each have an additional project that contains just an extension method for registering it with DI,

right. I think for e.g. AppInsights, we'll have another extension methods that configure all OTel for AppInsights usage, would be great to have common look and feel for different exporters though. Stackdriver might want to choose this path too. 

> I don't know of how to fix that off the top of my head. It would need some investigation. There is probably some nice way to improve it though.

AppInsights work the same problem around with making something else (that will get resolved) depend on auto-collectors. I hope there is nice and a more explicit way too.

**Comment by cijothomas** [comment]
For Asp.Net Core, ApplicationInsights registers an `IStartupFilter`, and attempts to resolve config. This IStartupFilter is added to DI, and hence its guaranteed to run before `Configure` method. This way autocollectors need no be manually resolved.

https://github.com/microsoft/ApplicationInsights-aspnetcore/blob/develop/src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsStartupFilter.cs#L15

**Comment by lmolkova** [comment]
I believe this is fixed with #253
